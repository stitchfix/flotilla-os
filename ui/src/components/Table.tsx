import * as React from "react"
import { HTMLTable } from "@blueprintjs/core"
import { isEmpty, isArray } from "lodash"
import SortableTh from "./SortableTh"
import { SortOrder } from "../types"

type Column = {
  displayName: string
  render: (item: any) => React.ReactNode
  isSortable: boolean
}

type Props<T> = {
  items: T[]
  columns: { [key: string]: Column }
  getItemKey: (item: T, index: number) => any
  updateSort: (sortKey: string) => void
  currentSortKey: string
  currentSortOrder: SortOrder
}

class Table<T> extends React.Component<Props<T>> {
  render() {
    const {
      columns,
      items,
      getItemKey,
      updateSort,
      currentSortKey,
      currentSortOrder,
    } = this.props

    return (
      <HTMLTable striped bordered>
        <thead>
          <tr>
            {Object.entries(columns).map(([k, v]) => (
              <SortableTh
                isSortable={v.isSortable}
                isActive={currentSortKey === k}
                order={currentSortOrder}
                onClick={updateSort.bind(this, k)}
                key={k}
              >
                {v.displayName}
              </SortableTh>
            ))}
          </tr>
        </thead>
        <tbody>
          {isArray(items) &&
            !isEmpty(items) &&
            items.map((item, i) => (
              <tr key={getItemKey(item, i)}>
                {Object.entries(columns).map(([k, v]) => (
                  <td key={k}>{v.render(item)}</td>
                ))}
              </tr>
            ))}
        </tbody>
      </HTMLTable>
    )
  }
}

export default Table
