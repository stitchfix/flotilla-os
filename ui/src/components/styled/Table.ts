import styled, { css } from "styled-components"
import colors from "../../helpers/colors"
import { DEFAULT_BORDER } from "../../helpers/styles"

export const Table = styled.table`
  background: ${colors.black[0]};
  display: flex;
  flex-flow: column nowrap;
  overflow: hidden;
  width: 100%;
`

export const TableRow = styled.tr`
  align-items: center;
  background: inherit;
  border-bottom: 1px solid ${colors.black[3]};
  display: flex;
  flex-flow: row nowrap;
  height: 48px;
  justify-content: flex-start;
  max-width: 100%;
  width: 100%;

  &:hover {
    background: ${colors.black[2]};
  }
`

const cellStyles = css`
  align-items: center;
  display: flex;
  flex-flow: row nowrap;
  justify-content: flex-start;
  min-width: 0;
  overflow: hidden;
  padding: 0 8px;
  text-overflow: ellipsis;
  white-space: nowrap;
  height: 100%;
  /* border-left: ${DEFAULT_BORDER}; */
`

export const TableCell = styled.td`
  ${cellStyles};
  flex: ${({ width }: { width?: number }) => (!!width ? width : 1)};
`

export const TableHeaderSortIcon = styled.div`
  margin-left: 4px;
  font-size: 0.6rem;
`

export const TableHeaderCell = styled.th`
  ${cellStyles};
  font-size: 0.9rem;
  text-transform: uppercase;
  font-weight: 500;
  border-top: none;
  flex: ${({ width }: { width?: number }) => (!!width ? width : 1)};
`

export const TableHeaderCellSortable =
  styled(TableHeaderCell) <
  { isActive: boolean, width: number } >
  `
  cursor: pointer;
  color: ${({ isActive }: { isActive?: boolean }) =>
    isActive ? colors.blue[0] : colors.gray[4]};
  background: ${({ isActive }: { isActive?: boolean }) =>
    isActive ? colors.black[1] : colors.black[0]};

  &:hover {
    color: ${colors.blue[0]};
    background: ${colors.black[1]};
  }
`
