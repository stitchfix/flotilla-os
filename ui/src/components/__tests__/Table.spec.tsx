import React from "react"
import { mount, ReactWrapper } from "enzyme"
import { mountToJson } from "enzyme-to-json"
import Table from "../Table"

describe("Table", () => {
  type MockItem = {
    id: number
    name: string
  }
  let wrapper: ReactWrapper
  const updateSort = jest.fn()

  beforeEach(() => {
    wrapper = mount(
      <Table<MockItem>
        items={[
          { id: 1, name: "One" },
          { id: 2, name: "Two" },
          { id: 3, name: "Three" },
        ]}
        getItemKey={(i: MockItem) => i.id}
        updateSort={updateSort}
        columns={{
          id: {
            displayName: "ID",
            render: (item: MockItem) => item.id,
            isSortable: true,
          },
          name: {
            displayName: "Name",
            render: (item: MockItem) => item.name,
            isSortable: true,
          },
        }}
      />
    )
  })

  it("renders", () => {
    expect(mountToJson(wrapper)).toMatchSnapshot()
  })

  it("calls props.updateSort when a sortable column header is clicked", () => {})
})
