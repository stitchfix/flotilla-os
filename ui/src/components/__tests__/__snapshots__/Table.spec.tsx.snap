// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders 1`] = `
<Table
  columns={
    Object {
      "id": Object {
        "displayName": "ID",
        "isSortable": true,
        "render": [Function],
      },
      "name": Object {
        "displayName": "Name",
        "isSortable": true,
        "render": [Function],
      },
    }
  }
  getItemKey={[Function]}
  items={
    Array [
      Object {
        "id": 1,
        "name": "One",
      },
      Object {
        "id": 2,
        "name": "Two",
      },
      Object {
        "id": 3,
        "name": "Three",
      },
    ]
  }
  updateSort={[MockFunction]}
>
  <HTMLTable
    bordered={true}
    striped={true}
  >
    <table
      className="bp3-html-table bp3-html-table-bordered bp3-html-table-striped"
    >
      <thead>
        <tr>
          <Th
            isActive={false}
            isSortable={true}
            key="id"
            onClick={[Function]}
          >
            <th
              className="flotilla-th-sortable"
              onClick={[Function]}
            >
              ID
            </th>
          </Th>
          <Th
            isActive={false}
            isSortable={true}
            key="name"
            onClick={[Function]}
          >
            <th
              className="flotilla-th-sortable"
              onClick={[Function]}
            >
              Name
            </th>
          </Th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="1"
        >
          <td
            key="id"
          >
            1
          </td>
          <td
            key="name"
          >
            One
          </td>
        </tr>
        <tr
          key="2"
        >
          <td
            key="id"
          >
            2
          </td>
          <td
            key="name"
          >
            Two
          </td>
        </tr>
        <tr
          key="3"
        >
          <td
            key="id"
          >
            3
          </td>
          <td
            key="name"
          >
            Three
          </td>
        </tr>
      </tbody>
    </table>
  </HTMLTable>
</Table>
`;
